### PR TITLE
Mining manager

### PR DESCRIPTION
**Pull request by tmpolaczyk**
Depends on #251

### CODE REVIEW

File: core/src/actors/mining_manager/handlers.rs [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+                    System::current().registry().get::<ReputationManager>();
+                reputation_manager_addr
+                    .send(ValidatePoE {
+                        beacon,
+                        proof: leadership_proof,
+                    })
+                    .into_actor(act)
+                    .drop_err()
+                    .and_then(move |eligible, _act, _ctx| {
+                        if eligible {
```
Comment (aesedepece) [inline core/src/actors/mining_manager/handlers.rs]: Let's log this as `info`.

File: config/src/config/mod.rs [hunk 2]
```diff
@@ -124,6 +127,13 @@ pub struct JsonRPC {
     pub server_address: SocketAddr,
 }
 
+/// JSON-RPC API configuration
```
Comment (aesedepece) [inline config/src/config/mod.rs]: This comment seems not to belong here.

File: config/src/config/mod.rs [hunk 3]
```diff
@@ -71,6 +71,9 @@ pub struct Config {
 
     /// JSON-RPC API configuration
     pub jsonrpc: JsonRPC,
+
+    /// MiningManager configuration
```
Comment (mariocao) [inline config/src/config/mod.rs]: Following previous comments, I would suggest:
```
/// Mining-related configuration
```

File: core/src/actors/mining_manager/actor.rs [hunk 4]
```diff
@@ -0,0 +1,16 @@
+use actix::{Actor, Context};
+
+use crate::actors::config_manager::send_get_config_request;
+
+use super::MiningManager;
+
+/// Make actor from EpochManager
```
Comment (mariocao) [inline core/src/actors/mining_manager/actor.rs]:
```suggestion
/// Make actor from MiningManager
```

File: core/src/actors/mining_manager/handlers.rs [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+////////////////////////////////////////////////////////////////////////////////////////
+// ACTOR MESSAGE HANDLERS
+////////////////////////////////////////////////////////////////////////////////////////
+
+/// Payload for the notification for all epochs
+#[derive(Clone, Debug)]
+pub struct EveryEpochPayload;
+
+/// Handler for EpochNotification<EveryEpochPayload>
+impl Handler<EpochNotification<EveryEpochPayload>> for MiningManager {
```
Comment (aesedepece) [inline core/src/actors/mining_manager/handlers.rs]: This handler's logic is too long. Let's see what can be taken out into static functions.

File: core/src/actors/mining_manager/handlers.rs [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+                    // We got a block from the future
+                    error!("The current highest checkpoint beacon is from the future");
+                    return actix::fut::err(());
+                }
+                if beacon.checkpoint == msg.checkpoint {
+                    // For some reason we already got a valid block for this epoch
+                    // TODO: Check eligibility anyway?
+                }
+                // TODO: this is correct, right?
+                beacon.checkpoint = msg.checkpoint;
```
Comment (aesedepece) [inline core/src/actors/mining_manager/handlers.rs]: This is right as long as we update the block digest of the `HighestCheckpointBeacon` in `ChainManager` at some point.

File: core/src/actors/mining_manager/mod.rs [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+use self::handlers::EveryEpochPayload;
+
+mod actor;
+mod handlers;
+
+////////////////////////////////////////////////////////////////////////////////////////
+// ACTOR BASIC STRUCTURE
+////////////////////////////////////////////////////////////////////////////////////////
+/// EpochManager actor
```
Comment (mariocao) [inline core/src/actors/mining_manager/mod.rs]:
```suggestion
/// MiningManager actor
```

File: docs/architecture/managers/mining_manager.md [hunk 8]
```diff
@@ -0,0 +1,117 @@
+<<<<<<< HEAD
```
Comment (mariocao) [inline docs/architecture/managers/mining_manager.md]: It seems you had some conflicts in here. :stuck_out_tongue:

Comment (tmpolaczyk) [reply docs/architecture/managers/mining_manager.md]: Shame on me :see_no_evil:

File: core/src/actors/mining_manager/handlers.rs (hunk 1 above)
Comment (tmpolaczyk) [reply core/src/actors/mining_manager/handlers.rs]: `info!("Mined a block!");` would be enough or do we want to log the proof of eligibility?

Comment (aesedepece) [reply core/src/actors/mining_manager/handlers.rs]: Something similar to this should do the job:
```rust
info!("Discovered eligibility for mining a block for epoch #{:?}");
```

File: core/src/actors/mining_manager/handlers.rs [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+            .then(move |beacon_msg, act, ctx| {
+                // Check eligibility
+                // S(H(beacon))
+                let mut beacon = match beacon_msg {
+                    Ok(Ok(b)) => b,
+                    _ => return actix::fut::err(()),
+                };
+                if beacon.checkpoint > msg.checkpoint {
+                    // We got a block from the future
+                    error!("The current highest checkpoint beacon is from the future");
```
Comment (aesedepece) [inline core/src/actors/mining_manager/handlers.rs]:
```suggestion
                    error!("The current highest checkpoint beacon is from the future ({:?} > {:?})", beacon.checkpoint, msg.checkpoint);
```

**Pull request merged**